Documentation overhaul, linting, and file transfer optimizations

yianzhang14 opened the pull request:
- overhauled documentation throughout the codebase, adding explanatory comments, function headers, and more typing
- added linting with ESLint configurations & updates to the typescript compiler
- added more robust null checking throughout the codebase
- created a new revised documentation markdown file
- added caching for file transfers to HPCs
- added manifest-specific extraction from git repos instead of having to repull everything
- removed unnecessary packages in the package.json

yianzhang14 committed:
refactored statistic

yianzhang14 committed:
Merge remote-tracking branch 'upstream/v2' into v2

yianzhang14 committed:
refactoring + documentation

yianzhang14 committed:
finished refactoring on maintainers

yianzhang14 committed:
finished connector refactoring

yianzhang14 committed:
finished eslinting all remaining ts files

yianzhang14 committed:
documented server.ts + added markdown files

yianzhang14 committed:
refactored cli.ts

yianzhang14 committed:
fixed some typing issues

yianzhang14 committed:
revamped packages + redid express structure + more linting

yianzhang14 committed:
more type hinting + async removal on models

yianzhang14 committed:
linting on main src/ files

yianzhang14 committed:
changes to type assertions

yianzhang14 committed:
finished maintainers

yianzhang14 committed:
make types more meaningful

yianzhang14 committed:
finished with util rework + minor changes elsewhere

yianzhang14 committed:
finished linting + helper rework + better nullguarding + max line size

yianzhang14 committed:
fixed helper typo & slurmconnector doc

yianzhang14 committed:
add specialized manifest download functionality + lint action

yianzhang14 committed:
documentation initial commit + helper moving + abstract functions + initial lazy supervisor

yianzhang14 committed:
finished codebase overview doc

yianzhang14 committed:
finished codebase documentation doc

yianzhang14 committed:
added caching to folder uploaders + a bit more linting

yianzhang14 committed:
disable cached upload on data

yianzhang14 committed:
added cache refresing functionality

yianzhang14 committed:
fixed swagger error

yianzhang14 committed:
fixed cli

yianzhang14 committed:
readd typedoc

yianzhang14 committed:
more package updates

yianzhang14 committed:
added mysql

yianzhang14 committed:
disable uml

yianzhang14 committed:
move all documentation out of build

yianzhang14 committed:
fixed generate swagger bug

yianzhang14 committed:
reenable documentation

yianzhang14 committed:
fixed gitutil bug + updated packages

yianzhang14 committed:
recursive mkdirs

yianzhang14 committed:
more gitutil bugfixes

yianzhang14 committed:
silenced console.logs

yianzhang14 committed:
reenabled urlencoded

yianzhang14 committed:
fixed folder uploader bug

yianzhang14 committed:
more bugfixing + removing console logs

yianzhang14 committed:
Merge branch 'v2' of https://github.com/yianzhang14/cybergis-compute-core-refactor into v2

yianzhang14 committed:
fixed remotefsexists bug

yianzhang14 committed:
job upload now works

yianzhang14 committed:
update build script to copy example jsons

yianzhang14 committed:
fixed github actions

yianzhang14 commented:
Fixed some remaining bugs and did more rigorous testing: 

- passed the sample test bench at https://github.com/cybergis/cybergis-compute-api-tests
- submitted a sample job, which went through the pipeline successfully

also fixed the github actions to copy config jsons and to actually lint things

yianzhang14 committed:
added ruff for the python files for now

yianzhang14 committed:
added typing to cachedupload

yianzhang14 committed:
switched to git sdk

yianzhang14 committed:
switch to fs promises

yianzhang14 committed:
automated branch resolution + out of date checking

yianzhang14 committed:
moved configs to example

yianzhang14 committed:
undo config.json move

yianzhang14 committed:
fixed action typo

yianzhang14 committed:
removed all refreshing code + added automatic logic to decide when to re-upload git repo

yianzhang14 committed:
added yaml parser

yianzhang14 committed:
rework folder typing to be more enumy

yianzhang14 committed:
leftover changes from folder rework

yianzhang14 committed:
added cache model

yianzhang14 committed:
force synchronize

yianzhang14 committed:
synchronize in the json

yianzhang14 committed:
actually force reinitialize

yianzhang14 committed:
turned strict on + updated packages + migrated to typeorm 0.3 + migrated to redis 5 + cleaned up redis classes

yianzhang14 committed:
reworked tsconfig

yianzhang14 committed:
remove typeorm-uml

yianzhang14 committed:
actually disable uml + add cache to entities

yianzhang14 committed:
actually disable uml instead of tsdoc

yianzhang14 commented:
- Upgraded redis & typeorm to more recent versions
- Cleaned up all redis cache classes (don't need to promisify anymore -- redis v5 supports promises natively)
- added a cache table for automatic caching functionality
- added a git sdk and ported all git calls to it (also used for resolving branch names and determining commit divergence)

yianzhang14 committed:
fixed redis

yianzhang14 committed:
connect to redis client

yianzhang14 committed:
reworked cache + deprecated specialized manifest + added more git caching

yianzhang14 committed:
reworked refreshing in getexecutablemanifest

yianzhang14 committed:
reworked db to not be a class per typeorm guidelines

yianzhang14 committed:
leftover db port changes

yianzhang14 committed:
remove ormconfig

yianzhang14 committed:
initialize database in server.ts

yianzhang14 committed:
fixed some minor errors

yianzhang14 committed:
initialize datasource in init hello world

yianzhang14 committed:
reenable register

yianzhang14 committed:
move this.register

yianzhang14 committed:
Merge branch 'v2' of https://github.com/yianzhang14/cybergis-compute-core-refactor into v2

yianzhang14 committed:
fix db registering

yianzhang14 committed:
fix db registering

yianzhang14 committed:
Merge branch 'v2' of https://github.com/yianzhang14/cybergis-compute-core-refactor into v2

yianzhang14 committed:
add print statements

yianzhang14 committed:
debugging job2object

yianzhang14 committed:
updated dockerfile + removed print statements

yianzhang14 committed:
fixed typing on nullable model columns

yianzhang14 committed:
better undefined/null handling + typeorm function fixes

yianzhang14 committed:
added more is_testing clauses to console logs

yianzhang14 committed:
fixed registercache

yianzhang14 committed:
more fixes to hpcpath -> cachepath

yianzhang14 committed:
disable removing comments

yianzhang14 committed:
Merge branch 'v2' of https://github.com/yianzhang14/cybergis-compute-core-refactor into v2

yianzhang14 committed:
styling change + git manifest cache + new git routes

yianzhang14 committed:
rework server structure

yianzhang14 committed:
initialize cache

yianzhang14 committed:
change rimraf

yianzhang14 committed:
changed promises to awaits

yianzhang14 committed:
change unix time scaling

yianzhang14 committed:
fix linting in cli

alexandermichels commented on FolderRouter.ts:
```diff
@@ -0,0 +1,349 @@
+import express = require("express");
```
Are these the exact same routes just split into multiple files?

yianzhang14 replied:
yes

yianzhang14 replied:
although with some added routes, as discussed

alexandermichels commented on configs/examples/config.example.json:
This change needs to be undone as it will break existing scripts that copy the example JSONs into normal JSONs.

yianzhang14 replied:
Reverted

alexandermichels reviewed (dismissed):
Is it possible to split this PR between the linting/doc changes and your actual code changes? Or at least could you let me know which files have actual changes to the functionality vs. just linting?

yianzhang14 committed:
undo examples folder

yianzhang14 commented:
> Is it possible to split this PR between the linting/doc changes and your actual code changes? Or at least could you let me know which files have actual changes to the functionality vs. just linting?

I could try to separate the PR out, but I think that would take a considerable amount of time since everything is kinda intertwined at this point. However, I think all the changes can be boiled down into four categories:

1. **Pure linting changes** - these come from ESLint & are largely stylistic/have little semantic changes on what the code does (aside from stuff like changing vars to consts, which required some restructuring of things, especially in for loops--iterateindices to values w/ for-each loops). This also includes various comments/docstrings I added throughout the codebase and me changing all ==/!= to more safe === !== (which could break things if the unsafeness was intended). As a result, these changes largely should be non-breaking.
2. **Typescript compilation changes** - these come from setting strict to true in the tsconfig.json file. Resolving these took a bit more effort/changes beyond just code style. I think the main problems I had to fix from this involved getting rid of all Anys, adding explicit typing everywhere I could, moving state up and down the inheritance hierarchy, changing some interface type signatures from ?: to : (i.e., force them to not be undefined/null), and adding explicit assertions that things aren't null/undefined when we expect them not to be. These changes also should be largely unbreaking, since they only affect/relate to the compilation of things, and not their actual runtime behavior.
3.  **Library upgrades/changes** - these come from updating libraries to newer versions and removing unneeded libraries. For the former, this largely involves updating typeorm and redis and adding the isomorphic-git sdk. With typeorm 0.3, the API underwent a significant change, so `DB.ts` was altered significantly, and through this, all database interactions had their syntax changed (but this was just a direct translation of the previous things). With the new redis versions, things just became a lot easier to use, and I essentially just took all the scattered redis classes defined throughout and centralized them into an inheritance hierarchy within `Redis.ts`. With isomorphic-git, I replaced most of the `execs` in `GitUtil.ts`. For the removing packages part, this was relatively straightfoward as these packages were not used anywhere or were used sparsely--e.g., bodyparser for express, which was originally a one-liner middleware for parsing request jsons but which is now deprecated (native express can do this now). 
4. **Actual rewrites** - these actually change the logic of the code and either implement new features/make convoluted things more streamlined. The two major sources of this includes `GitUtil.ts` and `FolderUploader.ts`, which were changed to allow for manifest caching and server-side git repo caching.

yianzhang14 commented:
As a brief outline of which files fall into what category:

- Essentially everything falls into categories 1/2, as all files needed restructuring/syntax changes to conform to `strict: true` and the eslint guidelines. I also added comments/docstrings essentially everywhere in the core code files. This is also the bulk of the raw changes in this PR. 
- For category 3, the `typeorm` upgrade involves all the files in the `models/` folder and the `DB.ts` file, but every line that accesses the database throughout also had minor changes due to API/syntax changes with the new version. The `redis` upgrade mainly involves `Redis.ts`, which centralizes all the original classes that accessed redis in the codebase (which were kinda just scattered throughout--e.g., credentialmanager and queue). The isomorphic-git changes mainly involve `GitUtil.ts`, replacing `execs`. 
- The bulk of the category 4 changes revolve around `FolderUploader.ts`, `GitUtil.ts`, and `models/Cache.ts` to add the functionality to enable caching. There are a few more changes of this type scattered sparsely throughout the codebase iirc, but all I remember about them is that I only did these when the original code was egregiously convoluted and could be simplified significantly, which did not occur often.

alexandermichels reviewed (changes requested):
Two big things now:

1. There are conflicts that block from merging because of another PR that was merged, please resolve those conflicts.
2. I'm unable to deploy the code on our test server. I updated the image in our compute_stack.yml and deleted node_modules, but when I deploy, I get the following error:

```
> cybergis-compute-core@1.0.0 swagger
> node ./production/tools/generate-swagger.js

supervisor server is up, listening to port: 3030
Error during Data Source initialization: QueryFailedError: Cannot enqueue Query after invoking quit.
    at Query.<anonymous> (/job_supervisor/node_modules/typeorm/driver/mysql/MysqlQueryRunner.js:165:37)
    at Query.<anonymous> (/job_supervisor/node_modules/mysql/lib/Connection.js:526:10)
    at Query._callback (/job_supervisor/node_modules/mysql/lib/Connection.js:488:16)
    at Sequence.end (/job_supervisor/node_modules/mysql/lib/protocol/sequences/Sequence.js:83:24)
    at /job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:236:14
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  query: 'ALTER TABLE `logs` CHANGE `deletedAt` `deletedAt` bigint(6) NULL',
  parameters: undefined,
  driverError: Error: Cannot enqueue Query after invoking quit.
      at Protocol._validateEnqueue (/job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:215:16)
      at Protocol._enqueue (/job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:138:13)
      at PoolConnection.query (/job_supervisor/node_modules/mysql/lib/Connection.js:198:25)
      at /job_supervisor/node_modules/typeorm/driver/mysql/MysqlQueryRunner.js:154:36
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 'PROTOCOL_ENQUEUE_AFTER_QUIT',
    fatal: false
  },
  code: 'PROTOCOL_ENQUEUE_AFTER_QUIT',
  fatal: false
}
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

QueryFailedError: Cannot enqueue Query after invoking quit.
    at Query.<anonymous> (/job_supervisor/node_modules/typeorm/driver/mysql/MysqlQueryRunner.js:165:37)
    at Query.<anonymous> (/job_supervisor/node_modules/mysql/lib/Connection.js:526:10)
    at Query._callback (/job_supervisor/node_modules/mysql/lib/Connection.js:488:16)
    at Sequence.end (/job_supervisor/node_modules/mysql/lib/protocol/sequences/Sequence.js:83:24)
    at /job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:236:14
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  query: 'ALTER TABLE `logs` CHANGE `deletedAt` `deletedAt` bigint(6) NULL',
  parameters: undefined,
  driverError: Error: Cannot enqueue Query after invoking quit.
      at Protocol._validateEnqueue (/job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:215:16)
      at Protocol._enqueue (/job_supervisor/node_modules/mysql/lib/protocol/Protocol.js:138:13)
      at PoolConnection.query (/job_supervisor/node_modules/mysql/lib/Connection.js:198:25)
      at /job_supervisor/node_modules/typeorm/driver/mysql/MysqlQueryRunner.js:154:36
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: 'PROTOCOL_ENQUEUE_AFTER_QUIT',
    fatal: false
  },
  code: 'PROTOCOL_ENQUEUE_AFTER_QUIT',
  fatal: false
}

Node.js v18.18.0
```

You can poke around on the test server to figure out what's happening, but remember that if we test and approve we will need to deploy to production as well.

yianzhang14 commented:
I'll start working on these in the next few days. For the second issue, I think that may be due to me changing some database schemas to comply with the typescript compiler. I'll try to revert those with some exception comment/flag--they weren't exactly critical.

alexandermichels commented:
Sounds good. You don't necessarily have to revert the database changes if you have a simple solution for updating an existing database to fit the new schema. So for example, a script that can update the data types/whatever needs to be done. Just need a way to push these changes to dev for testing and eventually production. Great work so far!

alexandermichels commented:
@yianzhang14 we are trying to include this in next month's release, so we need this sorted out soon and deployed on our test server. When do you think you can get this working with instructions for deploying to test/production?

yianzhang14 commented:
Sorry, just recently got back to the US but I should have some time this week to sort it out. Can you let me know the exact server you're trying to deploy to?

alexandermichels commented:
No worries. The exact server is less important than having the documented process for deploying it. We will have to deploy this to test, then production and since this is OSS we should have documentation to deploy anywhere. 

You can try on test and keep track of any changes/code you run. I'll use that documentation to deploy on production.

yianzhang14 committed:
rename a file

yianzhang14 committed:
Merge remote-tracking branch 'upstream/v2' into v2

yianzhang14 committed:
fixed merge bugs

yianzhang14 committed:
fixed linting

yianzhang14 committed:
change the eslint config

yianzhang14 committed:
revert back to commonjs

yianzhang14 committed:
change global variables

yianzhang14 commented:
I just fixed the merge conflicts, but I can't seem to recreate the error you received with typeorm. The errors it's throwing shouldn't be changing anything, as I only changed some type annotations.

alexandermichels commented:
Yeah not sure what happened, maybe it just needed to be run twice to reformat the DB? Everything is working with the new commits on test. 

One issue when running the `three-examples` job: I saw this issue in the logs:

```
Assertion failed: 'Variable is undefined/null when it should not be. Assertion at Error, SingularityConnector.prepare: 70'
Assertion failed: 'Variable is undefined/null when it should not be. Assertion at Error, SingularityConnector.prepare: 71'
```
however, the job still ran fine. May just want to look into that to see if its an issue or not. Going to ask some collaborators to test out their more complex jobs on the server now.

yianzhang14 commented:
I think I brought that up in some meeting before. I essentially added some sanity checks that print those things if things were null, and that popped up for some user-related things when building the slurm config. It doesn't seem to affect anything from what I've seen, but it should probably be resolved at some point.
